lightningd/plugin.c

```diff
@@ -1146,3 +1161,51 @@ struct log *plugin_get_log(struct plugin *plugin)
 {
 	return plugin->log;
 }
+
+struct command_result *json_reload_plugins(struct command *cmd,
+						const char *buffer,
+						const jsmntok_t *obj UNNEEDED,
+						const jsmntok_t *params)
+{
+	const char *plugin_path;
+	struct plugin *p;
+	struct json_stream *response;
+
+	if (!param(cmd, buffer, params,
+		   p_opt("directory", param_string, &plugin_path),
+		   NULL))
+		return command_param_failed();
+
+	if (plugin_path) {
+		DIR *d = opendir(plugin_path);
+		if (!d)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					"Could not open %s", plugin_path);
+		closedir(d);
+		add_plugin_dir(cmd->ld->plugins, plugin_path, true);
+	} else
+		plugins_add_default_dir(cmd->ld->plugins,
+					  path_join(tmpctx, cmd->ld->config_dir, "plugins"));
+	plugins_init(cmd->ld->plugins, cmd->ld->dev_debug_subprocess);
```

Comment:
`plugin_init` clears the `pending_requests` field, which is used to route plugin responses back to the command (I think).  So if we execute `reload_plugin` while a plugin command is ongoing, this will prevent those command responses from being recognized one the plugin finishes the command, and also potentially leak the memory involved.

Should `reload_plugin` actively *kill* running plugins (and possibly raise an `error` on existing commands that are being handled by the plugin)?

Reply:
> plugin_init clears the pending_requests field, which is used to route plugin responses back to the command (I think). So if we execute reload_plugin while a plugin command is ongoing, this will prevent those command responses from being recognized once the plugin finishes the command, and also potentially leak the memory involved.

>

Reply:
> I thought that nobody would reload plugins while a command while pending..

Nobody might do so deliberately, but there exist such things as bugs.

Reply:
> I cannot see yet any code that kills plugins for reload_plugins, is this deliberate or did I miss?

Actually it only loads plugin which have not been loaded at launch. The name might not be accurate..

Reply:
`load_new_plugins`?

If so then running plugin-provided commands should be unaffected. Suggest refactoring the clearing of the fields away from starting new plugins then.